flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java

```diff
@@ -39,12 +40,29 @@ public interface LeaderElectionEventHandler {
      */
     void onGrantLeadership(UUID newLeaderSessionId);
 
+    /**
+     * Called by specific {@link LeaderElectionDriver} when the leadership is granted.
+     *
+     * <p>This method will trigger the grant event processing in a separate thread.
+     *
+     * @param newLeaderSessionId the valid leader session id
+     */
+    CompletableFuture<Void> onGrantLeadershipAsync(UUID newLeaderSessionId);
```

Comment:
What's the long-term plan here; will we keep both variants and everything has to support both? Or will we drop the sync variants once we remove legacy stuff?

Reply:
In the long run, we could remove the async calls again from the interface. The driver implementations only need the synchronous methods only because their event handling lives in a separate thread already. The asynchronous execution would become an implementation detail of `DefaultLeaderElectionService`

Reply:
Considering that, adding deprecation annotations here would make sense.

Reply:
Why can't it be an implementation detail of the `DefaultLeaderElectionService` _now_?

Reply:
How would we determine that we're running in the main thread? We would need to get a reference to the owner of `DefaultLeaderElectionService`, wouldn't we? That way we would be able to check whether the call is run in the main thread of the owner. But this would closer coupling of the owner and the `DefaultLeaderElectionService`. Or do you have something else in mind which I'm missing? :thinking: 

My understanding is that we have the following requirements:
* `onGrantLeadership`, `onRevokeLeadership` and `onLeaderInformationChanged` should be called in a single thread to ensure sequential execution.
* the methods shouldn't be called in the main thread of the `LeaderElectionService`'s owner. The k8s and ZK implementation do this implicitly through their event thread handling. The `DefaultMultipleComponentLeaderElectionService` doesn't do that (anymore). But that's an implementation detail of the service. That's why I thought that it's also the `DefaultMultipleComponentLeaderElectionService`'s responsibility to call the event processing functions asynchronously.

WDYT?

Reply:
> We would need to get a reference to the owner of `DefaultLeaderElectionService`, wouldn't we?

Ok, I was not thinking straight when writing the previous comment. We have the `LeaderContender`'s reference. 🤦 I'm gonna give it some thought.

Reply:
> How would we determine that we're running in the main thread?

Do we need to know that?